[lldb-dap] Migrating DAP 'initialize' to new typed RequestHandler.

ashgti opened the pull request:
This adds new types and helpers to support the 'initialize' request with the new typed RequestHandler. While working on this I found there were a few cases where we incorrectly treated initialize arguments as capabilities. The new `lldb_dap::protocol::InitializeRequestArguments` and `lldb_dap::protocol::Capabilities` uncovered the inconsistencies.

ashgti committed:
[lldb-dap] Migrating DAP 'initialize' to new typed RequestHandler.

This adds new types and helpers to support the 'initialize' request with the new typed RequestHandler. While working on this I found there were a few cases where we incorrectly treated initialize arguments as capabilities. The new `lldb_dap::protocol::InitializeRequestArguments` and `lldb_dap::protocol::Capabilities` uncovered the inconsistencies.

labath reviewed (commented):
This code seems quite repetitive. Would it be possible to have a `Capabilities` enum and then representing the actual values by something like `map<Capability, bool>` (with std::nullopt replaced by removing the capability from the map). Or even `map<string, bool>` if you don't need to access the capabilities individually (very often)? And maybe the handlers don't even need to return the map, but just a list/vector of "new" capabilities that they enable?

ashgti committed:
Converting the std::optional<bool> values in Capabilities and IntializeRequestHandler into a `std::set<Feature>` to simplify these flags.

ashgti commented:
> This code seems quite repetitive. Would it be possible to have a `Capabilities` enum and then representing the actual values by something like `map<Capability, bool>` (with std::nullopt replaced by removing the capability from the map). Or even `map<string, bool>` if you don't need to access the capabilities individually (very often)? And maybe the handlers don't even need to return the map, but just a list/vector of "new" capabilities that they enable?

Done. Converted these into `std::set<(Adapter|Client)Feature>` sets for the `Capabilities` type and for the `InitializeRequestArguments`. I adjusted the RequestHandler classes to return a `std::set<AdapterFeature>` now as well.

JDevlieghere commented on lldb/tools/lldb-dap/Protocol/ProtocolRequests.h:
```diff
@@ -54,6 +54,79 @@ bool fromJSON(const llvm::json::Value &, DisconnectArguments &,
+  /// If true all column numbers are 1-based (default).
+  std::optional<bool> columnsStartAt1;
+
+  enum class Feature {
+    /// Client supports the `type` attribute for variables.
+    supportsVariableType,
+    /// Client supports the paging of variables.
+    supportsVariablePaging,
+    /// Client supports the `runInTerminal` request.
+    supportsRunInTerminalRequest,
+    /// Client supports memory references.
+    supportsMemoryReferences,
+    /// Client supports progress reporting.
+    supportsProgressReporting,
+    /// Client supports the `invalidated` event.
+    supportsInvalidatedEvent,
+    /// Client supports the `memory` event.
+    supportsMemoryEvent,
+    /// Client supports the `argsCanBeInterpretedByShell` attribute on the
+    /// `runInTerminal` request.
+    supportsArgsCanBeInterpretedByShell,
+    /// Client supports the `startDebugging` request.
+    supportsStartDebuggingRequest,
+    /// The client will interpret ANSI escape sequences in the display of
+    /// `OutputEvent.output` and `Variable.value` fields when
+    /// `Capabilities.supportsANSIStyling` is also enabled.
+    supportsANSIStyling,
+  };
```
I know this is matching the spec, but I think this looks really weird and the more we move in the protocol class, the more inconsistent the code becomes and that's a bad thing. It stands out but not in a good way as the casing of the spec isn't meaningful. IMHO we shouldn't match the casing of the spec and remained consistent with LLDB.

ashgti replied:
I adjusted the naming to match how other lldb enums are defined. This is now `enum ClientFeature` with keys like `eClientFeatureSupportsStartDebuggingRequest`. I can shorten those if you want me to drop say the `supports` part of the enum label.

JDevlieghere replied:
I'm fine either way. Looking at the spec, all the boolean capabilities start with Supports so I think it's fine to drop that part.

ashgti replied:
I dropped 'support' from the name to shorten things slightly.

JDevlieghere commented on lldb/tools/lldb-dap/Protocol/ProtocolTypes.cpp:
```diff
@@ -44,4 +44,223 @@ bool fromJSON(const json::Value &Params, Source &S, json::Path P) {
+  json::Object result;
+
+  for (const auto &feature : C.supportedFeatures)
+    switch (feature) {
+    case Capabilities::Feature::supportsANSIStyling:
+      result.insert({"supportsANSIStyling", true});
+      break;
```
I would create a (static) helper that converts the enum values to strings and then do 
```result.insert({GetFeatureAsString(feature), true})';```

ashgti replied:
Done.

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -1144,31 +1145,45 @@ lldb::SBValue Variables::FindVariable(uint64_t variablesReference,
-  capabilities["supportsLoadedSourcesRequest"] = false;
-  capabilities["supportsRestartFrame"] = false;
-  capabilities["supportsStepBack"] = false;
+  // supportsGotoTargetsRequest, supportsLoadedSourcesRequest,
+  // supportsRestartFrame, supportsStepBack
```
Looking at the spec, it seems like we can just omit unsupported capabilities. Should we just remove this then?

ashgti replied:
Removed.

JDevlieghere commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -47,7 +48,7 @@ class BaseRequestHandler {
   virtual void operator()(const protocol::Request &request) const = 0;
 
-  virtual llvm::StringMap<bool> GetCapabilities() const { return {}; }
+  virtual std::set<AdapterFeature> GetSupportedFeatures() const { return {}; }
```
Should this return an `llvm::SmallSet<AdapterFeature,1>?`? And then merge them into a `llvm::DenseSet`?

ashgti replied:
Done.

JDevlieghere commented on lldb/tools/lldb-dap/Protocol/ProtocolRequests.cpp:
```diff
@@ -23,6 +24,75 @@ bool fromJSON(const json::Value &Params, DisconnectArguments &DA,
+    return false;
+
+  const json::Object *O = Params.getAsObject();
+  if (std::optional<bool> v = O->getBoolean("supportsVariableType"); v && *v)
+    IRA.supportedFeatures.insert(ClientFeature::supportsVariableType);
```
This seems like an opportunity for a for-loop that iterates over a `std::pair<lllvm::StringLiteral, ClientFeature>`.

ashgti replied:
Added a lookup table for these key/value pairs.

JDevlieghere commented on lldb/tools/lldb-dap/Protocol/ProtocolRequests.h:
```diff
@@ -54,6 +54,79 @@ bool fromJSON(const llvm::json::Value &, DisconnectArguments &,
+  std::optional<bool> columnsStartAt1;
+
+  enum class Feature {
+    /// Client supports the `type` attribute for variables.
+    supportsVariableType,
+    /// Client supports the paging of variables.
+    supportsVariablePaging,
+    /// Client supports the `runInTerminal` request.
+    supportsRunInTerminalRequest,
+    /// Client supports memory references.
+    supportsMemoryReferences,
+    /// Client supports progress reporting.
+    supportsProgressReporting,
+    /// Client supports the `invalidated` event.
+    supportsInvalidatedEvent,
+    /// Client supports the `memory` event.
```
I don't mind re-adding those comments, but I omitted them in my earlier patch because they're pretty much adding no value. I think `supportsMemoryEvent` and `supportsANSIStyling` are the only two that deserve a comment.

ashgti replied:
I trimmed most of the comments.

JDevlieghere commented on lldb/tools/lldb-dap/Protocol/ProtocolTypes.h:
```diff
@@ -23,10 +23,240 @@
+
+/// Names of checksum algorithms that may be supported by a debug adapter.
+/// Values: ‘MD5’, ‘SHA1’, ‘SHA256’, ‘timestamp’.
+enum class ChecksumAlgorithm { md5, sha1, sha256, timestamp };
```
Shouldn't these match the spec? 
```suggestion
enum class ChecksumAlgorithm { MD5, SHA1, SHA256, timestamp };
```

ashgti committed:
Apply suggestions from code review

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

ashgti committed:
Adjusting enum naming to match lldb styles, converting from set to DenseSet/SmallDenseSet and using some lookup tables to map strings to enum values.

JDevlieghere commented on lldb/tools/lldb-dap/Handler/RequestHandler.h:
```diff
@@ -47,7 +49,9 @@ class BaseRequestHandler {
   virtual void operator()(const protocol::Request &request) const = 0;
 
-  virtual llvm::StringMap<bool> GetCapabilities() const { return {}; }
+  virtual llvm::SmallDenseSet<AdapterFeature, 1> GetSupportedFeatures() const {
```
Let's use a typedef to make this a little less verbose. 
```suggestion
  using FeatureSet =  llvm::SmallDenseSet<AdapterFeature, 1>;
  virtual FeatureSet GetSupportedFeatures() const {
```

JDevlieghere commented on lldb/tools/lldb-dap/Protocol/ProtocolTypes.cpp:
```diff
@@ -44,4 +43,190 @@ bool fromJSON(const json::Value &Params, Source &S, json::Path P) {
+  switch (feature) {
+  case eAdapterFeatureSupportsANSIStyling:
+    return "supportsANSIStyling";
+  case eAdapterFeatureSupportsBreakpointLocationsRequest:
+    return "supportsBreakpointLocationsRequest";
```
I feel bad for not thinking about this earlier, but now that I see this switch, I think this would' bea good candidate for a `.def` file. We could have a `dap.def` file that defines and use that to implement methods for both serializing and deserializing. That doesn't need to be part of this PR though.

JDevlieghere reviewed (approved):
LGTM

ashgti committed:
Update lldb/tools/lldb-dap/Handler/RequestHandler.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

ashgti committed:
Shortening the name of feature flags to remove 'support'.

JDevlieghere reviewed (approved):
Thanks, I'm happy with this!

ashgti committed:
Adjusting whitespace in a comment.
